bytecomp/matching.ml

```diff
@@ -3101,6 +3223,51 @@ let rec compile_match repr partial ctx (m : pattern_matching) =
       (bind_check str v arg lam, total)
   | _ -> assert false
 
+and compile_simplified repr partial ctx
+    (m : Simple.clause pattern_matching) =
+  match m with
+  | { cases = []; args = [] } -> comp_exit ctx m
+  | { args = ((Lvar v as arg), str) :: argl } ->
+      (* Not sure we need to do anything with arg/v *)
+      let first_match, rem =
+        split_and_precompile_simplified { m with args = (arg, Alias) :: argl }
+      in
+      let lam, total =
+        comp_match_handlers
+          (( if dbg then
+             do_compile_matching_pr
+           else
+             do_compile_matching
+           )
+             repr)
+          partial ctx first_match rem
+      in
+      (bind_check str v arg lam, total)
+  | _ -> assert false
+
+and compile_half_compiled repr partial ctx
+    (m : Typedtree.pattern Non_empty_clause.t pattern_matching) =
+  match m with
+  | { cases = []; args = [] } -> comp_exit ctx m
+  | { args = ((Lvar v as arg), str) :: argl } ->
+      (* Not sure we need to do anything with arg/v *)
+      let first_match, rem =
+        split_and_precompile_nonempty (Some v)
+          { m with args = (arg, Alias) :: argl }
+      in
+      let lam, total =
+        comp_match_handlers
+          (( if dbg then
+             do_compile_matching_pr
+           else
+             do_compile_matching
+           )
+             repr)
+          partial ctx first_match rem
+      in
+      (bind_check str v arg lam, total)
+  | _ -> assert false
```

Comment:
I'm slightly annoyed by the fact that this exists in 3 versions.
Any suggestion regarding how to dedup that?